src/Shared/PackageManagers/NPMProjectManager.cs

```diff
@@ -74,6 +75,24 @@ public override async Task<IEnumerable<string>> DownloadVersion(PackageURL purl,
             return downloadedPaths;
         }
 
+        /// <summary>
+        /// Check if the package exists in the respository.
+        /// </summary>
+        /// <param name="purl">The PackageURL to check.</param>
+        /// <param name="useCache">If cache should be used.</param>
+        /// <returns>True if the package is confirmed to exist in the repository. False otherwise.</returns>
+        public override async Task<bool> PackageExists(PackageURL purl, bool useCache = true)
+        {
+            Logger.Trace("PackageExists {0}", purl?.ToString());
+            if (purl is null || purl.Name is null || purl.Type is null)
+            {
+                Logger.Trace("Provided PackageURL was null.");
+                return false;
+            }
```

Comment:
It seems that this code is repeated in every `PackageManager`. Could we refactor it somewhere to like a helper method or the method implementation or something? Just to avoid repetition.

Reply:
Unfortunately I think if we refactor this out we lose the ability to trace the nullability and later to satisfy the compiler would need to handle purl remaining nullable (even though the helper method should confirm it is not nullable.) 

Perhaps it could be worked around with a method like:

bool PurlIsValid(PackageURL purl, out PackageURL notNullPurl)

But you can also note that it's not *exactly* the same for each because some managers care about, for example; the namespace field, and some do not.